### PR TITLE
svchost/auth: New API for storing (rather than just reading) credentials

### PR DESCRIPTION
**Pull request by apparentlymart**
Our initial `CredentialsSource` idea only implemented a read operation for retrieving credentials from either CLI configuration or a helper program. Here we extend that to now also support storing and then later forgetting credentials.

There is no caller for this new API yet, but it will follow in a later commit as we implement the new `terraform login` command. I implemented the new methods for both local file storage (creating a `credentials.tf.json` file in the CLI configuration automatically) and for writing to a credentials helper program using some new verbs.

Our initial implementation of `CredentialSource` was built by composing together a number of separate types to create a layer-based credentials store. However, with the new write calls in place, the resulting code became pretty hard to follow with everything split apart like that, so this also includes a new `CredentialsSource` implementation that includes all of the source-selection logic in one place as regular sequential code, so it's (subjectively) easier to see how things all fit together.

The new `CredentialSource` implementation _will_ be used for reads once this is merged, even though the store/forget parts of it are dead code for the moment pending `terraform login` implementation.

---

The CLI config is still being parsed using HCL 1 right now, and there are no short-term plans to change that, but since I was introducing some new API surface here anyway I pre-emptively switched the low-level credentials representation to be `cty.Value`, currently shimmed from HCL 1 values using `hcl2shim`, in preparation for a future change to make the CLI config use HCL 2.

### CODE REVIEW

File: command/cliconfig/config_windows.go [hunk 1]
```diff
@@ -44,3 +47,24 @@ func homeDir() (string, error) {
...
+	srcPtr, err := syscall.UTF16PtrFromString(source)
+	if err != nil {
+		return &os.LinkError{"replace", source, destination, err}
+	}
+	destPtr, err := syscall.UTF16PtrFromString(destination)
+	if err != nil {
+		return &os.LinkError{"replace", source, destination, err}
+	}
+
+	err := windows.MoveFileEx(srcPtr, destPtr, flags)
```
Comment (jbardin) [inline command/cliconfig/config_windows.go]: you have a re-declared error here (windows file, so not tested directly)

Comment (apparentlymart) [reply command/cliconfig/config_windows.go]: Good catch! Thanks.

Comment (apparentlymart) [reply command/cliconfig/config_windows.go]: Okay... peeking at my local history, I see that I over-zealously simplified this function after I'd initially tested it, reorganizing this code so that it was all in one function rather than two, but then apparently neglected to actually test the resulting simplification. :confounded: 

With this code now split out into its own PR I don't have a manual testing codepath to run this against right now, so I've just verified that it's _buildable_ on Windows now and I'll retest it working on Windows once this is in and I've got `terraform login` running against it.
